### ISSUE TITLE
Sequence pad support for Ragged Tensor when call to_tensor

### ISSUE THREAD

**Issue by workingloong** [tensorflow/tensorflow#34793]
**System information**
- TensorFlow version (you are using): 2.0.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Right now, we can get a dense tensor by calling `to_tensor` of a RaggedTensor and the num_cols of the dense tensor is the max length of the tensors in the RaggedTensor. However, the max length may vary if we get the RaggedTensor by `tf.string.splits` for different batch data. So, we can not use the dense tensor for `tf.keras.layers.Embedding` and `tf.keras.layers.Dense` layers because the num_cols is not fixed.

**Will this change the current api? How?**
In order to support padding for `to_tensor`, we need add three arguments for `to_tensor` API, like:
```python
def to_tensor(self, default_value=None, name=None, 
    maxlen=None, padding='pre', truncating='pre'):
```
maxlen: Int, maximum length of all sequences where the sequence is a tensor in the RaggedTensor .
padding: String, 'pre' or 'post':
    pad either before or after each sequence.
truncating: String, 'pre' or 'post':
    remove values from sequences larger than
    `maxlen`, either at the beginning or at the end of the sequences.

Those arguments are the same as `tf.keras.preprocessing.sequence.pad_sequences` in https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/sequence/pad_sequences

**Who will benefit with this feature?**
In NLP and time series modeling, the format of feature is the string with many values separated by a separator like "19,39,94,30" or "there, is, a, dog" and the number of values of different samples is different. In this case, we need to use `tf.strings.split()` to get a RaggedTensor and convert it to a dense tensor with fixed num_cols.

**Any Other info.**
Here is a demo with maxlen.
```python
import numpy as np

from tensorflow.python import tf2
from tensorflow.python.client import session
from tensorflow.python.framework import composite_tensor
from tensorflow.python.framework import constant_op
from tensorflow.python.framework import dtypes
from tensorflow.python.framework import ops
from tensorflow.python.framework import sparse_tensor
from tensorflow.python.framework import tensor_shape
from tensorflow.python.framework import tensor_spec
from tensorflow.python.framework import tensor_util
from tensorflow.python.framework import type_spec
from tensorflow.python.ops import array_ops
from tensorflow.python.ops import check_ops
from tensorflow.python.ops import control_flow_ops
from tensorflow.python.ops import gen_ragged_conversion_ops
from tensorflow.python.ops import math_ops
from tensorflow.python.ops.ragged import ragged_config
from tensorflow.python.ops.ragged import ragged_tensor_value
from tensorflow.python.ops.ragged import ragged_util
from tensorflow.python.ops.ragged import segment_id_ops
from tensorflow.python.util.tf_export import tf_export

def is_ragged(value):
  """Returns true if `value` is a ragged tensor or ragged tensor value."""
  return isinstance(value,
                    (tf.RaggedTensor, ragged_tensor_value.RaggedTensorValue))

def to_tensor(ragged_tensor, default_value=None, name=None, maxlen=None):
    self = ragged_tensor
    print("self : ", self)
    with ops.name_scope(name, "RaggedToTensor", [self, default_value]):
      if default_value is not None:
        default_value = ops.convert_to_tensor(
            default_value, name="default_value", dtype=self.dtype)

      # If ragged_rank > 1, then recursively convert the ragged values into a
      # `Tensor` before we proceed.
      values = self.values
      if is_ragged(values):
        values = values.to_tensor(default_value)

      # Tile the default value, if necessary.
      if default_value is not None:
        if values.shape.ndims is not None:
          default_value.shape.with_rank_at_most(values.shape.ndims - 1)
        if (values.shape.ndims is None or default_value.shape.ndims is None or
            values.shape.ndims != default_value.shape.ndims + 1):
          value_shape = array_ops.shape(values)[1:]
          default_value = array_ops.broadcast_to(default_value, value_shape)
        default_value.shape.assert_is_compatible_with(values.shape[1:])

      # Get the expected dense shape ([nrows, ncols] + value_shape).
      rt_row_lengths = [self.row_splits[1:] - self.row_splits[:-1]]
      print('rt_row_lengths : ', rt_row_lengths)
      nrows = array_ops.shape(self.row_splits,
                              out_type=self._row_splits.dtype)[0] - 1
      ncols = math_ops.maximum(math_ops.reduce_max(rt_row_lengths), 0) 
      if maxlen is not None:
        ncols = tf.constant(maxlen, dtype=tf.int64)
        
      values_shape = array_ops.shape(values, out_type=self._row_splits.dtype)
      value_shape = values_shape[1:]
      nvals = values_shape[0]

      # Build a default value if none was supplied.
      if default_value is None:
        default_value = array_ops.zeros(value_shape, dtype=values.dtype)
      default_value.shape.assert_is_compatible_with(values.shape[1:])
      default_value.set_shape(values.shape[1:])

      # Get the row start indices, and expand to shape=[nrows, 1].
      starts = array_ops.expand_dims(self.row_splits[:-1], 1)

      # Get the row limit indices, and expand to shape=[nrows, 1].
      limits = array_ops.expand_dims(self.row_splits[1:], 1)

      # Get the column indices, and expand to shape=[1, ncols].
      columns = array_ops.expand_dims(math_ops.range(0, ncols), 0)

      # Build a list containing the values plus the default value.  We will use
      # tf.gather to collect values from this list for the `Tensor` (using
      # nvals as the index for the default value).
      values_and_default = array_ops.concat(
          [values, array_ops.stack([default_value])], axis=0)

      # Construct a matrix "indices" pointing into values_and_default.  I.e.,
      # output[r, c] = values_and_default[indices[r, c].
      nondefault_index = starts + columns
      has_value = nondefault_index < limits
      default_index = array_ops.fill(array_ops.stack([nrows, ncols]), nvals)
        
      indices = array_ops.where(has_value, nondefault_index, default_index)
      tensor = array_ops.gather(values_and_default, indices)

      # Gather the results into a `Tensor`.
      return tensor
    
rt = tf.ragged.constant([[9, 8, 7], [], [6, 5], [4]])
print(to_tensor(rt, maxlen=10))
```

**Comment by workingloong** [comment]
@gadagashwini @ymodak  Can you assign the issue to me? I have implemented the feature and test it.

**Comment by edloper** [comment]
RaggedTensor.to_tensor has a shape parameter (as of 0efc9f4) which you can use to pad/truncate to a specified shape. Partial shapes (ie shapes where some dimensions have unknown size) are supported.

**Comment by workingloong** [comment]
Good, Thank!  Can you add a padding argument which we can use to pad either before or after each sequence? So we can pad the ragged tensor like `tf.keras.preprocessing.sequence.pad_sequences` https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/sequence/pad_sequences.
For example:
```
>>> rt = tf.ragged.constant([[9, 8, 7], [], [6, 5], [4]])
>>> print(rt.to_tensor(shape=(4,3), padding="pre"))
[[9 8 7] [0 0 0] [6 5 0] [4 0 0]], shape=(4, 3), dtype=int32)

>>> print(rt.to_tensor(shape=(4,3), padding="post"))
[[9 8 7] [0 0 0] [0 6 5] [0 0 4]], shape=(4, 3), dtype=int32)
```

**Comment by anantvir** [comment]
This thread was useful but seems incomplete. Can someone please help me out here.

I also have text data(sentences of varying length). Can I use to_tensor to pad each sentence to a max length of 200 words ?

**Comment by edloper** [comment]
@anantvir If you have a 2D ragged tensor `sentences` with shape `[num_sentences, tokens_per_sentence]`, and you want to pad it exactly 200 tokens in each row, you can just use `sentences.to_tokens(shape=[None, 200])`.  E.g.:

```python
>>> rt = tf.ragged.constant([[1, 2], [], [3, 4, 5, 6]])
>>> print(rt.to_tensor(shape=[None, 200]))
tf.Tensor(
[[1 2 0 0 0 0 ...]
 [0 0 0 0 0 0 ...]
 [3 4 5 6 0 0 ...]], shape=(3, 200), dtype=int32)
```

If you want to pad it to have at most 200 tokens in each row (but less if all sentences are shorter than 200 tokens), then you could use `sentences[:, :200].to_tensor()`.  E.g.:

```
>>> print(rt[:, :200].to_tensor())
tf.Tensor(
[[1 2 0 0]
 [0 0 0 0]
 [3 4 5 6]], shape=(3, 4), dtype=int32)
```

@workingloong If you want to put padding before each row (rather than after), then you can't currently do that with `RaggedTensor.to_tensor`.  But you can write a fairly simple function to do it:

```
>>> def left_pad_2d_ragged(rt, width):
...   rt = rt[-width:]  # Truncate rows to have at most `width` items
...   pad_row_lengths = width - rt.row_lengths()
...   pad_values = tf.zeros([(width * rt.nrows()) - tf.size(rt, tf.int64)], rt.dtype)
...   padding = tf.RaggedTensor.from_row_lengths(pad_values, pad_row_lengths)
...   return tf.concat([padding, rt], axis=1).to_tensor()
>>> print(left_pad_2d_ragged(rt, 10))
tf.Tensor(
[[0 0 0 0 0 0 0 0 1 2]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 3 4 5 6]], shape=(3, 10), dtype=int32)
```

If you'd be interested in putting together a PR that updates `to_tensor` to support pre/post padding, I'd be happy to take a look.  But I'd want it to be sufficiently general -- e.g., if there are multiple ragged dimensions, then we should probably be able to specify whether each dimension should have adding placed at the beginning or end of each slice.
